sdks/go/pkg/beam/runners/prism/internal/engine/strategy.go

```diff
@@ -23,16 +23,351 @@ import (
+		sub.reset(state)
+	}
+	delete(state.Trigger, t)
+}
+
+// TriggerAfterEach processes each trigger before executing the next.
+// Starting with the first subtrigger, ready when the _current_ subtrigger
+// is ready. After output, advances the current trigger by one.
+type TriggerAfterEach struct {
+	SubTriggers []Trigger
+}
+
+func (t *TriggerAfterEach) isReady(input triggerInput, state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+	// Use extra for the current trigger index.
+	if ts.extra == nil {
+		ts.extra = int(0)
+	}
+	current := ts.extra.(int)
+	ready := false
+	if t.SubTriggers[current].isReady(input, state) {
+		current++
+		ready = true
+	}
+	if current >= len(t.SubTriggers) {
+		ts.finished = true
+	}
+	ts.extra = current
+	state.setTriggerState(t, ts)
+	return ready
+}
+
+func (t *TriggerAfterEach) reset(state *StateData) {
+	for _, sub := range t.SubTriggers {
+		sub.reset(state)
+	}
+	delete(state.Trigger, t)
+}
+
+// TriggerElementCount triggers when there have been at least the required number
+// of elements have arrived.
+type TriggerElementCount struct {
+	ElementCount int
+}
+
+func (t *TriggerElementCount) isReady(input triggerInput, state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+
+	if ts.extra == nil {
+		ts.extra = int(0)
+	}
+	count := ts.extra.(int) + input.newElementCount
+	ts.extra = count
+	ready := count >= t.ElementCount
+	if ready {
+		ts.finished = true
+	}
+	state.setTriggerState(t, ts)
+	return ready
+}
+
+func (t *TriggerElementCount) reset(state *StateData) {
+	delete(state.Trigger, t)
+}
+
+// TriggerOrFinally is ready whenever either of it's subtriggers fire.
+// Ceases to be ready after the Finally trigger isReady.
+type TriggerOrFinally struct {
+	Main    Trigger // repeated
+	Finally Trigger // terminates execution.
+}
+
+func (t *TriggerOrFinally) isReady(input triggerInput, state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+
+	mainReady := t.Main.isReady(input, state)
+	finallyReady := t.Finally.isReady(input, state)
+
+	if mainReady || finallyReady {
+		t.Main.reset(state)
+	}
+
+	if finallyReady {
+		t.Finally.reset(state)
+		ts.finished = true
+	}
+
+	state.setTriggerState(t, ts)
+	return mainReady || finallyReady
+}
+
+func (t *TriggerOrFinally) reset(state *StateData) {
+	t.Main.reset(state)
+	t.Finally.reset(state)
+	delete(state.Trigger, t)
+}
+
+// TriggerRepeatedly is a composite trigger that will fire whenever the Repeated trigger is ready.
+// If the Repeated trigger is finished, it's state will be reset.
+type TriggerRepeatedly struct {
+	Repeated Trigger
+}
+
+func (t *TriggerRepeatedly) isReady(input triggerInput, state *StateData) bool {
+	mainReady := t.Repeated.isReady(input, state)
+
+	// If the subtrigger is finished, reset it.
+	if repeatedTs := state.getTriggerState(t.Repeated); repeatedTs.finished {
+		t.Repeated.reset(state)
```

Comment:
At this point I'm repeating myself but just to add one more example, this reset should happen in `onFire`